src/driver/mod.rs

```diff
@@ -0,0 +1,293 @@
+        }
+    }
+}
+
+/// <command name> <key> <flags> <exptime> <bytes> [noreply]\r\n
+///
+///
+/// - "STORED\r\n", to indicate success.
+///
+/// - "NOT_STORED\r\n" to indicate the data was not stored, but not
+/// because of an error. This normally means that the
+/// condition for an "add" or a "replace" command wasn't met.
+///
+/// - "EXISTS\r\n" to indicate that the item you are trying to store with
+/// a "cas" command has been modified since you last fetched it.
+///
+/// - "NOT_FOUND\r\n" to indicate that the item you are trying to store
+/// with a "cas" command did not exist.
+pub async fn storage<K>(
+    mut conn: PoolConnection<'_>,
+    command: StorageCommand,
+    key: K,
+    flags: u32,
+    expiration: impl Into<Option<Duration>>,
+    bytes: Vec<u8>,
+    noreply: bool,
+) -> Result<Response, MemcacheError>
+where
+    K: AsRef<[u8]>,
+{
+    // <command name>
+    let _ = conn.write(command.into()).await?;
+    // <key>
+    let _ = conn.write_all(key.as_ref()).await?;
+    let _ = conn.write(EMPTY_SPACE_BYTES).await?;
+
+    // <flags>
+    let _ = conn.write(flags.to_string().as_ref()).await?;
+    let _ = conn.write(EMPTY_SPACE_BYTES).await?;
+
+    // <exptime>
+    let exptime = expiration.into().map(|d| d.as_secs()).unwrap_or(0);
+    let _ = conn.write(exptime.to_string().as_ref()).await?;
+    let _ = conn.write(EMPTY_SPACE_BYTES).await?;
+
+    // <bytes>
+    let _ = conn.write(bytes.len().to_string().as_bytes()).await?;
+
+    // [noreply]
+    if noreply {
+        // FYI: NO_REPLY_BYTES contains space before and new line after
+        let _ = conn.write(NO_REPLY_BYTES).await?;
+    } else {
+        let _ = conn.write(NEW_LINE_BYTES).await?;
+    }
+
+    // <data block>
+    let _ = conn.write_all(&bytes).await?;
+    let _ = conn.write(NEW_LINE_BYTES).await?;
+
+    // Flush command
+    let _ = conn.flush().await?;
+
+    // 64 bytes should be enough to address all storage responses
+    let mut buffer: BytesMut = BytesMut::with_capacity(64);
+
+    let _ = conn.read_buf(&mut buffer).await?;
+
+    match parser::parse_ascii_status(&buffer) {
+        Ok((_left, result)) => Ok(result),
+        Err(e) => Err(MemcacheError::Nom(format!("{}", e))),
+    }
+}
+
+/// Retrieval command
+#[derive(Debug)]
+pub enum RetrievalCommand {
+    /// "get" means "get this data".
+    Get,
+    /// "gets" means "get multiple data".
+    Gets,
+}
+
+impl From<RetrievalCommand> for &'static [u8] {
+    fn from(c: RetrievalCommand) -> &'static [u8] {
+        match c {
+            RetrievalCommand::Get => b"get",
+            RetrievalCommand::Gets => b"gets",
+        }
+    }
+}
+
+/// get <key>*\r\n
+/// gets <key>*\r\n
+///
+///
+/// VALUE <key> <flags> <bytes> [<cas unique>]\r\n
+/// <data block>\r\n
+/// VALUE <key> <flags> <bytes> [<cas unique>]\r\n
+/// <data block>\r\n
+/// VALUE <key> <flags> <bytes> [<cas unique>]\r\n
+/// <data block>\r\n
+/// "END\r\n"
+pub async fn retrieve<K>(
```

Comment:
Will this always return single or multiple values? If this might differ, I'd consider splitting this into two methods 🤔

Reply:
Yeah, I'd say this could be beneficial and more API-friendly to be able to get Vec<Value> and Option<Value>.